LibOS/shim/src/bookkeep/shim_vma.c

```diff
@@ -65,16 +73,23 @@ struct shim_vma {
  */
 static inline void * __malloc (size_t size)
 {
-    struct shim_thread * thread = get_cur_thread();
-
+    void * addr;
     size = ALIGN_UP(size);
-    void * addr = (void *) DkVirtualMemoryAlloc(NULL, size, 0,
-                                                PAL_PROT_WRITE|PAL_PROT_READ);
+
+    /*
+     * Chia-Che 3/3/18: We must enforce the policy that all VMAs have to
+     * be created before issuing the PAL calls.
+     */
+    addr = __bkeep_unmapped(PAL_CB(user_address.end),
+                            PAL_CB(user_address.start), size,
+                            PROT_READ|PROT_WRITE,
+                            MAP_PRIVATE|MAP_ANONYMOUS|VMA_INTERNAL,
+                            NULL, 0, "vma");
```

Comment:
Maybe this change will finally fix #100, let's hope so.

Reply:
I'll note that this is a good candidate for the type of synchronization policy that should probably be collected in one place, like a header file that articulates how to add more memory allocation interfaces the "right" way.

Reply:
Documented.